### PR TITLE
Add support for transactions

### LINKED ISSUE #83: InnoDB slowness
Creation/population of some tables is slow in InnoDB.
As some tables are re-created each run (set and grouping tables), these wull use the default MySQL storage engine. This used to be MyISAM, but is now InnoDB.
This can lead to  various engines being used e.g.
```
+--------------------------------+--------+
| TABLE_NAME                     | ENGINE |
+--------------------------------+--------+
| irstats2_browsers              | MyISAM |
| irstats2_cache_set_values      | InnoDB |
| irstats2_countries             | MyISAM |
| irstats2_deposits              | InnoDB |
| irstats2_doc_access            | InnoDB |
| irstats2_doc_format            | InnoDB |
| irstats2_downloads             | MyISAM |
| irstats2_groupings_authors     | InnoDB |
| irstats2_groupings_iau         | InnoDB |
| irstats2_groupings_institution | InnoDB |
| irstats2_groupings_type        | InnoDB |
| irstats2_history               | MyISAM |
| irstats2_internal              | MyISAM |
| irstats2_referrer              | MyISAM |
| irstats2_search_terms          | MyISAM |
| irstats2_sets_authors          | InnoDB |
| irstats2_sets_iau              | InnoDB |
| irstats2_sets_institution      | InnoDB |
| irstats2_sets_type             | InnoDB |
| irstats2_views                 | MyISAM |
+--------------------------------+--------+
```
The `irstats2_cache_set_values`, `irstats2_groupings_*`, and `irstats2_sets_*` tables do not have a primary key, so record insertion with the InnoDB engine is slower (http://stackoverflow.com/questions/9114209/innodb-inserts-very-slow-and-slowing-down).

### PR DESCRIPTION
**Pull request by jesusbagpuss**
Fixes #83

### CODE REVIEW

File: cfg/plugins/EPrints/Plugin/Stats/Handler.pm [hunk 1]
```diff
@@ -637,13 +637,17 @@ sub save_data_values
         {
                 foreach my $epid ( keys %{$data->{$date}} )
                 {
+						$self->{dbh}->begin; #start transaction if DB engine supports it
```
Comment (graingert) [inline cfg/plugins/EPrints/Plugin/Stats/Handler.pm]: Redundant comment

File: cfg/plugins/EPrints/Plugin/Stats/Sets.pm [hunk 2]
```diff
@@ -230,6 +230,8 @@ sub populate_tables
 
 		my $eprintid = $eprint->get_id;
 
+		$self->handler->{dbh}->begin; #make all changes in one transaction if DB engine supports them
```
Comment (graingert) [inline cfg/plugins/EPrints/Plugin/Stats/Sets.pm]: Redundant comment

File: cfg/plugins/EPrints/Plugin/Stats/Sets.pm [hunk 3]
```diff
@@ -257,6 +259,8 @@ sub populate_tables
 				$display_cache->{$set_name}->{$raw_value} = 1;
 			}
 		}
+		
+		$self->handler->{dbh}->commit; #end transaction
```
Comment (graingert) [inline cfg/plugins/EPrints/Plugin/Stats/Sets.pm]: Redundant comment

File: cfg/plugins/EPrints/Plugin/Stats/Handler.pm [hunk 4]
```diff
@@ -637,13 +637,17 @@ sub save_data_values
...
+						
                         foreach my $value ( keys %{$data->{$date}->{$epid}} )
                         {
                                 $i = 0;
                                 $sth->bind_param( ++$i, $_ ) for( ( $counter, $epid, $date, $value, ($data->{$date}->{$epid}->{$value}) ) );
                                 $rc &&= $sth->execute();
-				$counter++;
+                                $counter++;
```
Comment (graingert) [inline cfg/plugins/EPrints/Plugin/Stats/Handler.pm]: Unrelated whitespace changes

**Comment by graingert** [comment]
Any reason you're using begin over begin_work?

Is there a nice Perl dbi function that does automatic transactions/subtransactions with save points and manages exception rollbacks

**Comment by jesusbagpuss** [comment]
Hi @graingert - I tried `begin_work` when testing this - the method wasn't present.
Digging into the EPrints code, there are methods that exist - which I assumed were added for a reason (I'm not sure what reason) - so I used them...
